warehouse/accounts/services.py

```diff
@@ -759,3 +772,163 @@ def check_password(self, password, *, tags=None):
         # This service allows *every* password as a non-breached password. It will never
         # tell a user their password isn't good enough.
         return False
+
+
+class GitHubPublicKeyMetaAPIError(InvalidTokenLeakRequest):
+    pass
+
+
+@implementer(IGitHubTokenScanningPayloadVerifyService)
+class GitHubTokenScanningPayloadVerifyService:
+    """
+    Checks payload signature using:
+    - `requests` for HTTP calls
+    - `cryptography` for signature verification
+    """
+
+    def __init__(self, *, session, metrics, api_token=None):
+        self._metrics = metrics
+        self._session = session
+        self._api_token = api_token
+
+    @classmethod
+    def create_service(cls, context, request):
+        return cls(
+            session=request.http,
+            metrics=request.find_service(IMetricsService, context=context),
+            api_token=request.registry.settings["github.token"],
+        )
+
+    def verify(self, *, payload, key_id, signature):
+
+        try:
+            pubkey_api_data = self._retrieve_public_key_payload()
+            public_keys = self._extract_public_keys(pubkey_api_data)
+            public_key = self._check_public_key(
+                github_public_keys=public_keys, key_id=key_id
+            )
+            self._check_signature(
+                payload=payload, public_key=public_key, signature=signature
+            )
+        except InvalidTokenLeakRequest as exc:
+            self._metrics.increment(
+                f"warehouse.token_leak.github.auth.error.{exc.reason}"
+            )
+            return False
+
+        self._metrics.increment("warehouse.token_leak.github.auth.success")
+        return True
+
+    def _get_headers(self):
+        if self._api_token:
+            return {"Authorization": f"token {self._api_token}"}
+        return {}
+
+    def _retrieve_public_key_payload(self):
+        # TODO: cache ?
```

Comment:
I would agree, but honnestly I have no idea how often this would happen. Every minute ? Hour ? Day ? Week ?

Also, what would happen if someone purposefully (...or not?) posted millions of "things that look like a pypi token" ? If we check once per match, GitHub might end up thinking we're trying to DDOS them :)

Reply:
That's a fair point, we don't have a way to determine the request is legitimate before we get the public keys. That makes me think we probably _should_ cache this, which also should reduce the need for using a PAT with the request.

Reply:
Ok ! But I don’t know how caching works on Warehouse. Any pointer ?

Reply:
I don't believe we do anything like this at the moment unfortunately. This is (I think) the first time we're reaching out to a third-party service like this (that isn't on the frontend).

Perhaps just a simple in-memory TTL cache with https://pypi.org/project/cachetools/ would suffice?